network/src/protocols/discovery/mod.rs

```diff
@@ -357,18 +356,26 @@ where
                     assert_ne!(peer_id, self_peer_id);
                     // Update internal state of the peer with new Note.
                     self.known_peers.insert(peer_id, (peer_info.clone(), note));
+
+                    // The multiaddrs in the peer's discovery Note.
+                    let mut peer_addrs: Vec<Multiaddr> = peer_info
+                        .get_addrs()
+                        .iter()
+                        .cloned()
+                        .map(|addr| Multiaddr::try_from(addr).expect("Multiaddr parsing fails"))
+                        .collect();
+
+                    // Append the addrs in the seed PeerInfo if this peer is
+                    // configured as one of our seed peers.
+                    if let Some(seed_info) = self.seed_peers.get(&peer_id) {
+                        let seed_addrs_iter = seed_info.get_addrs().iter().cloned().map(|addr| {
+                            Multiaddr::try_from(addr).expect("Multiaddr parsing fails")
```

Comment:
expect results in a panic. This makes it easy for a remote validator to crash all other validators. We should not have expect used after the run time has started unless there is some truly unrecoverable behavior.

Reply:
@phlip9: definitely agree with the idea of separating unverified structs from verified ones -- [`bytecode_verifier`](https://github.com/libra/libra/blob/112cb1e7a5d81b90b191e2598620e3bcf1f4241a/language/bytecode_verifier/src/verifier.rs#L30-L34) uses type level state machines to encode invariants. Might be worth taking a look.